src/write_kosak.c

```diff
@@ -0,0 +1,2159 @@
+  bufprintf(ctx->buffer, ctx->size, "%s", (const char*)payload);
+}
+
+static void wg_json_uint64_handler(json_ctx_t *ctx, void *payload) {
+  bufprintf(ctx->buffer, ctx->size, "%" PRIu64, *(uint64_t*)payload);
+}
+
+static void wg_json_indent(json_ctx_t *ctx, int direction) {
+  int old_indent_level = ctx->indent_level;
+  ctx->indent_level += direction * ctx->indent_offset;
+  if (old_indent_level >= 0 && old_indent_level < sizeof(ctx->indent_buf) &&
+      ctx->indent_level >= 0 && ctx->indent_level < sizeof(ctx->indent_buf)) {
+    ctx->indent_buf[old_indent_level] = ' ';
+    ctx->indent_buf[ctx->indent_level] = 0;
+  }
+}
+
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// Context submodule. Defines the master wg_context_t object, which holds the
+// context for this plugin.
+//==============================================================================
+//==============================================================================
+//==============================================================================
+typedef struct {
+  server_ctx_t *server_ctx;
+  credential_ctx_t *cred_ctx;
+  oauth2_ctx_t *oauth2_ctx;
+  tree_ctx_t *tree_ctx;
+  pthread_mutex_t tree_mutex;
+} wg_context_t;
+
+static wg_context_t *wg_context_create(const char *project_id,
+                                       const char *instance_id,
+                                       const char *zone,
+                                       const char *email,
+                                       const char *key_file,
+                                       const char *passphrase);
+static void wg_context_destroy(wg_context_t *data);
+
+//------------------------------------------------------------------------------
+// Private implementation starts here.
+//------------------------------------------------------------------------------
+
+
+static wg_context_t *wg_context_create(const char *project_id,
+                                       const char *instance_id,
+                                       const char *zone,
+                                       const char *email,
+                                       const char *key_file,
+                                       const char *passphrase) {
+  wg_context_t *ctx = malloc_zero(sizeof(*ctx));
+  if (ctx == NULL) {
+    ERROR("wg_context_create: malloc_zero failed");
+    return NULL;
+  }
+
+  // Create the subcontext holding various pieces of server information.
+  ctx->server_ctx = wg_server_ctx_create(project_id, instance_id, zone);
+  if (ctx->server_ctx == NULL) {
+    ERROR("write_kosak: wg_server_ctx_create failed.");
+    wg_context_destroy(ctx);
+    return NULL;
+  }
+
+  // Optionally create the subcontext holding the service account credentials.
```

Comment:
We should probably also support the [Application Default Credentials](https://developers.google.com/identity/protocols/application-default-credentials) which allow all of this to be in a single file and also now there's a default location that it should be looked in (/etc/google/something_or_other... I'd have to go check in fluentd to see what salty is doing there)